src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+                            if nonsat == *k {
+                                self.stack.push(StackElement::Dissatisfied);
+                                None
+                            } else {
+                                Some(Err(Error::CouldnotEvaluate("Multisig dissatisfaction error".to_owned())))
+                            }
+                        }
+                        None => Some(Err(Error::CouldnotEvaluate("Unexpected Stack end".to_owned()))),
+                        _ => {
+                            self.state.push((root, NodeEvaluationState::PartialMultiCompleted(0, 0)));
+                            None
+                        }
+                    }
+                }
+            }
+            (&AstElem::ThreshM(ref k, ref subs), NodeEvaluationState::PartialMultiCompleted(i, n)) => {
+                if i == *k {
+                    if let Some(StackElement::Dissatisfied) = self.stack.pop(){
+                        self.stack.push(StackElement::Satisfied);
+                        None
+                    } else{
+                        Some(Err(Error::CouldnotEvaluate("Multisig extrazero not present".to_owned())))
+                    }
+                } else if n == subs.len(){
+                    Some(Err(Error::CouldnotEvaluate("MultiSig Error".to_owned())))
+                } else {
+                    match self.stack.evaluate_thres_m(self.secp, self.sighash, &subs[n]) {
+                        Some(Ok(x)) => {
+                            self.state.push((root, NodeEvaluationState::PartialMultiCompleted(i + 1, n + 1)));
+                            Some(Ok(x))
+                        }
+                        None => {
+                            self.state.push((root, NodeEvaluationState::PartialMultiCompleted(i, n + 1)));
+                            None
+                        }
+                        x => x //forward errors as is
+                    }
+                }
+            }
+            _ => Some(Err(Error::CouldnotEvaluate("Unexpected Interpretation error".to_owned())))
+        };
+        match node_res{
+            Some(res) => Some(res),
+            None => self.next()
+        }
+    }
+}
+
+/// Helper function to verify serialized signature
+pub fn verify_sersig(
+    secp: &secp256k1::Secp256k1<VerifyOnly>,
+    sighash: &secp256k1::Message,
+    pk: &bitcoin::PublicKey,
+    sigser: &[u8],
+) -> Result<secp256k1::Signature, Error>
+{
+    if let Some((_sighashtype, sig)) = sigser.split_last() {
+        match Signature::from_der(sig) {
+            Ok(sig ) => {
+                match secp.verify(&sighash, &sig, &pk.key) {
+                    Ok(()) => Ok(sig),
+                    Err(e) => Err(Error::Secp(e))
+                }
+            }
+            Err(e) => Err(Error::Secp(e))
+        }
+    }else{
+        return Err(Error::CouldnotEvaluate("Expected sighash type".to_owned()));
+    }
+}
+
+trait StackOperations<'stack>{
```

Comment:
It would probably be better to do
```
pub struct Stack(pub Vec<StackElement>)
```
and implement these methods directly on `Stack`, than to use this trait here.

Reply:
Do you want to propagate this all the way to the API inputs? Take input as `Stack` instead of `Vec<StackElement>`

Reply:
It's a bit ugly as I used functions of vector pop, push, len etc. Maybe I will redefine those again for this type

Reply:
Just use `.0` to get the inner vector.